### ISSUE TITLE
Reporting module error - mongodb binary object has no get

### ISSUE THREAD

**Issue by blk-ole** [kevoreilly/CAPEv2#1873]
## This is open source and you are getting __free__ support so be friendly!

# Prerequisites

Please answer the following questions for yourself before submitting an issue.

- [ x] I am running the latest version
- [ x] I did read the README!
- [ x] I checked the documentation and found no answer
- [ x] I checked to make sure that this issue has not already been filed
- [ x] I'm reporting the issue to the correct repository (for multi-repository projects)
- [ x] I have read and checked all configs (with all optional parts)

I'm looking for help with the following issue.

# Current Behavior

The analysis task is started and being run on the configured VM. The processing module reports no issue, but the task is never reported and not visible in the web view.

After looking into it, the MongoDB reporting module runs into the following error:

`

> 2023-12-01 09:39:24,172 [Task 5] [lib.cuckoo.core.plugins] DEBUG: Executing reporting module "MongoDB"
> 2023-12-01 09:39:24,184 [Task 5] [modules.reporting.mongodb] DEBUG: Deleted previous MongoDB data for Task 5
> 2023-12-01 09:39:24,321 [Task 5] [lib.cuckoo.core.plugins] ERROR: Failed to run the reporting module "MongoDB": 'Binary' object has no attribute 'get'
> Traceback (most recent call last):
>   File "/opt/CAPEv2/utils/../lib/cuckoo/core/plugins.py", line 715, in process
>     current.run(self.results)
>   File "/opt/CAPEv2/utils/../modules/reporting/mongodb.py", line 132, in run
>     mongo_insert_one("analysis", report)
>   File "/opt/CAPEv2/utils/../dev_utils/mongodb.py", line 55, in wrapper
>     return mongo_op_func(*args, **kwargs)
>   File "/opt/CAPEv2/utils/../dev_utils/mongodb.py", line 107, in mongo_insert_one
>     doc = hook(doc)
>   File "/opt/CAPEv2/utils/../dev_utils/mongodb.py", line 82, in inner
>     return f(*args, **kwargs)
>   File "/opt/CAPEv2/utils/../dev_utils/mongo_hooks.py", line 81, in normalize_files
>     for file_dict in collect_file_dicts(report):
>   File "/opt/CAPEv2/utils/../dev_utils/mongo_hooks.py", line 173, in collect_file_dicts
>     file_dicts.append(report.get("CAPE", {}).get("payloads", None) or [])
> AttributeError: 'Binary' object has no attribute 'get

'`

## Steps to Reproduce

Please provide detailed steps for reproducing the issue.

1. Install CapeV2 onto a fresh Ubuntu 22.04.1 installation via the kvm-qemu.sh and cape2.sh scripts. As written in the readme under "Installation recommendations [...]" and turning on the mongodb module + applying the asked for migrations.

OS: Ubuntu 22.04.1
CapeV2-Version: Installed on 02. November 2023
CapeV2 is installed in a virtual machine and the analysis VMs run as nested VMs. But this shouldn't cause this issue.

**Comment by doomedraven** [comment]
hello, can you share hash to try to reproduce that?

**Comment by blk-ole** [comment]
wow, that's a quick reply :D
many thanks!

Which hash do you mean?

**Comment by blk-ole** [comment]
Sha256: 0490ceace858ff7949b90ab4acf4867878815d2557089c179c9971b2dd0918b9
[0490ceace858ff7949b90ab4acf4867878815d2557089c179c9971b2dd0918b9.zip](https://github.com/kevoreilly/CAPEv2/files/13599000/0490ceace858ff7949b90ab4acf4867878815d2557089c179c9971b2dd0918b9.zip)

**Comment by doomedraven** [comment]
i can't reproduce your error so i have no idea what is wrong on your side

**Comment by blk-ole** [comment]
Thanks for looking into it.
I gonna retry with a fresh setup and report back

**Comment by doomedraven** [comment]
do you frequently doing `git pull` and restart of required services?

**Comment by blk-ole** [comment]
no not at all
the sandbox is still undergoing setup and has always been at git-commit

`/opt/CAPEv2$ git log
commit 3987ce591983e8e80ce97ccee156d490c0d565e8 (HEAD -> master, origin/master, origin/HEAD)
Author: doomedraven <[doommedraven@gmail.com](mailto:doommedraven@gmail.com)>
Date:   Fri Nov 24 09:05:12 2023 +0100

    sync`

**Comment by blk-ole** [comment]
sorry about that, didn't thought that the version would get old so fast.

**Comment by kevoreilly** [comment]
Bleeding edge sandbox 😎

**Comment by blk-ole** [comment]
Working bleeding edge! Something rare nowadays

**Comment by ethhart** [comment]
Looks like your retention time is too quick. It's deleting data before your processing finishes.
`[Task 5] [modules.reporting.mongodb] DEBUG: Deleted previous MongoDB data for Task 5`

**Comment by rtkcgrantcharov** [comment]
I've run into this issue as well. A regression was introduced when `dev_utils/mongodb_py` was split into:
- `dev_utils/mongodb_py`
- `dev_utils/mongo_hooks.py`

I reverted these 4 commits from my fork to get it working again:
- https://github.com/rtkwlf/sawndbox-cape/commit/f631f67bc04564cd582778eecdb3cc0b87c7207b
- https://github.com/rtkwlf/sawndbox-cape/commit/60bcb00b9c1a3adb84b99bf6e7dd0c00e3744395
- https://github.com/rtkwlf/sawndbox-cape/commit/9f7294211c0090c2f6b0f3a7a4fdfb9e9fbac075
- https://github.com/rtkwlf/sawndbox-cape/commit/0f420973eba2edaca3b9ae56f9f5e315ca3bf60b

The bug is in one of the above - 100% sure.

**Comment by rtkcgrantcharov** [comment]
Personally, I think a little bit more quality control on commits generally speaking would be nice ... finding a "working commit" to install off of is not fun. Neither is figuring out what to revert to make it work.

**Comment by kevoreilly** [comment]
Please take that attitude elsewhere. This is my life's work and reading comments like this after I've spent the last decade working hard on this project for free for others disgusts me quite frankly.

**Comment by rtkcgrantcharov** [comment]
Buddy, I didn't come here to throw shade on your life's work or act unappreciative of all the hard work you've put into this project. Take emotions aside for a second and focus on the objective facts.

1. If the above bug is a bug, it means that anybody who's tried to use your work for since Sept 19th hasn't been able to benefit from any of the subsequent value-add commits on new and updated signatures you've made since - simply because your project won't build correctly OR if it did build correctly, it then crapped out as soon as you ran an analysis on ANY piece of malware.
2. It's easy to sit on the sidelines and criticize someone else's work, but I'm not doing that ... I spent 2 hours debugging, compiling and testing a fix I then submitted into your cape2.sh logic 2 weeks ago because of change that appears went in entirely untested. I've personally done over 75+ builds of your code in the last 2 months.
3. No matter how good a project is there is ALWAYS room for improvement. You got 2 comments above saying "hey, we're having a hard time using this thing ... can we do something about it?" Not asking for more than a build job that runs once a week or something to make sure this thing can install successfully and run an analysis without crapping out along
 the way. In fact, one might already exist and I don't know about it ... and there may be a bug there.

I think this project is awesome, the work that is going into it is awesome and I'm simply asking for help from you guys to make it a bit more stable. Imagine this open source project was `git` and from one week to the next you lost the ability to commit code and it didn't get fixed for 3 months ... It's precisely because I find that what you guys are doing here valuable and awesome that I'm saying all this.

If you still find all this disgusting, I'm sorry to hear that ... but again ... no one came here to hurt anyones feelings or diminish/trivialize the work this project represents.

**Comment by doomedraven** [comment]
invoking me from PTO...

>>If the above bug is a bug, it means that anybody who's tried to use your work for since Sept 19th hasn't been able to benefit from any of the subsequent value-add commits on new and updated signatures you've made since - simply because your project won't build correctly OR if it did build correctly, it then crapped out as soon as you ran an analysis on ANY piece of malware. Isn't the fact that you guys didn't know about this until now a concern at all?

* That is so __FALSE__ dude, first go and see my initial answers, we can't reproduce that bug as I started to investigate that directly, we have a private slack with people from the world biggest IT companies, I personally own huge cluster on GCP, and 0 problems, running on latest always. So before coming and said that something is wrong and you need to search working commit is a bit funny, instead of search a working commit I would suggest you to review your setup as it sound mostly like you messed with it. Nobody from __advanced__ users have those problems, we tried to help to newcomer(who opened the issue), but if we can't reproduce it the same that issue doesn't exist aka user messed with his setup. 

*  About __value__ if you would be one of the advanced user you would see how much value it has. But each user case is different and is impossible to cover all cases.

TL;DR:  investigate your setup, the project works just fine if you have proper skills. Found what is wrong? provide as much details as you can to help us reproduce  and fix anything or add warning for those who mess due to improper usage.

Anyway detaching for the PTO again.

**Comment by doomedraven** [comment]
PD as curiosity I have dropped my mongodb, git clone master branch and did a fresh analysis of 1 tasks to not have any previous data in my cape to simulate fresh install. 0 issues, so for start of digging into your problems I would suggest to start doing diff of configs with repo and see what generate your problem and once again always try master before open issues

**Comment by rtkcgrantcharov** [comment]
First off, not my intention to detract from your PTO ... sorry that it has ...

I can appreciate you've tried to reproduce this bug, but just because haven't been successful doesn't mean it's not there. I reproduced it consistently at least 5 times with installations from vanilla Ubuntu 22.04.3 LTS with latest updates. I've also automated the entire install process as per the instructions listed here: https://capev2.readthedocs.io/en/latest/installation/index.html  using a combination of Terraform, Packer and Ansible so the "human error" is mostly out. 

I haven't done anything advanced other than run the following on a single VM:
 1. install `kvm-qemu.sh all cape`
 2. install `kvm-qemu.sh virtmanager`
 3. install `cape2.sh all cape`
 4. poetry install
 5. load 3 analysis vms with different versions of windows (w/ cape-agent installed)

I will give the "try latest master approach" another try if you're confident it's my setup and do another pass at verifying that I've followed all the steps in `readthedocs.io` ... in the process of automating things, I was very thorough ...

**Comment by doomedraven** [comment]
2 more things
I not saw mongodb version anywhere reported in issue
And another one the best way to deal with those cases is to put python
breakpoint adding it with try/except and reprocess the job and investigate
what is inside of the variable report

As we can't reproduce that + I'm on PTO I will leave this part to the
affected part, as you must understand if we can't reproduce it, there is
nothing that we can do

**Comment by doomedraven** [comment]
i would say your problem is here `compressresults` disable that and you should not suffer this, i personally want to nuke `compressresults` and `retention` modules. Do you know what `compressresults` does? it make your cape useless :D

**Comment by doomedraven** [comment]
confirmed

```
def run(self, results):

        for keyword in ("CAPE", "procdump"):
            if keyword in results:
                try:
                    cape_json = json.dumps(results[keyword], ensure_ascii=False).encode()
                    compressed_data = zlib.compress(cape_json)
                    results[keyword] = Binary(compressed_data) <- your problem here
```

so about "asking" for better quality commits. as you might see it not our problem, yes CAPE has a lot of modules. We don't use all of them, some are there as some users might have weird user cases or not understand what those modules does

and as you might see in configuration it marked as community module, which means we do not maintain this module
```
# Community
# Compress results including CAPE output
# to help avoid reaching the hard 16MB MongoDB limit.
[compressresults]
enabled = no
```

nuked compressresults from cape, `git pull` + `systemctl restart cape-processor` and you never will suffer this issue again.

So once again not just for you, but for everyone who comes over this, start with minimal config changes, if you have no clue what module does, __DISABLE__ it as it probably will give you just more problems than benefits

**Comment by ethhart** [comment]
Do we want to remove references to `compressresults` in [compare.py](https://github.com/kevoreilly/CAPEv2/blob/4da7e97c03a9e2aef08e4159f64278322bcf0484/lib/cuckoo/common/compare.py#L136C16-L136C16) and [analysis/views.py](https://github.com/kevoreilly/CAPEv2/blob/4da7e97c03a9e2aef08e4159f64278322bcf0484/web/analysis/views.py#L1394C32-L1394C32)?
I don't see any references in community to keep this logic in tact.

**Comment by rtkcgrantcharov** [comment]
Thanks for following up on this @doomedraven - at least we're all on the same page now.

Is it accurate to say that one of the 4 commits I referenced above introduced a side-effect bug in the `compressresults` feature? As I mentioned earlier, when I backed out the commits my setup resumed working as it had done before.

w.r.t.
> so about "asking" for better quality commits. as you might see it not our problem, yes CAPE has a lot of modules. We don't use all of them, some are there as some users might have weird user cases or not understand what those modules does

My asking for better quality commits is not entirely based on this bug - which was effectively a sizeable database change that impacted various parts of the codebase, had an unintended side-effect on an "unsupported feature" (i.e. `compressresults`), and went in with no evidence of any automated testing or anything of the sort.  Again, no clue what you guys do in the backend when vetting these changes (manual test?) ... the whole process is opaque from where I stand ... a lot of projects use **github-actions** automation for stuff like this. HOWEVER, my ask also comes on the heels of this fix that I put in a few weeks ago: https://github.com/kevoreilly/CAPEv2/pull/1850 ... you guys didn't catch it, I did, and I fixed it ... and I followed "used the latest master" branch which was broken at the time. Still took 2+ hours of trying to figure why my APT packages were all messed up and trace it up to a typo. We are all human, we all make mistakes ... not a big deal .. but things can be better with more testing and automation. That's all I'm going to say on this topic here as it's not related to this issue. I have 15+ years of dev/security experience from Fortune 500 companies and I given you my feedback - what you do with it from this point on is up to you.

**Comment by doomedraven** [comment]
well you can call it side effect if you want, but its a `community` module, as i told, we do not maintain those modules, so for us is not side effect, __is that community doesn't update their modules to align with core__. That would be a proper sentence.

We don't test and I personally don't care about maintaining community modules as it looks i need to say that lauder and harder. 

Dude i spend years doing automation of those scripts and any small change in 3rd part software breaks it, when i push changes to those scripts __is when im doing deployment and it works__, it can break in any moment and i mention that in README, that __IS NOT A SILVER BULLER__. 

now to the core msg of this all 
1. is software it changes = it might break
2. there is no paid support and done by people in their free time, spot bug, help us to fix, we can't cover all user cases
3. want a better core, help us to improve, that is more than welcome.
4. using community modules = it can be broken, it's not our problem is [mentioned in readme first line too](https://github.com/kevoreilly/CAPEv2/tree/master/conf#readme)

but all this just sound like you complain about open source project where nobody gets moneys for doing that and safe time to others, its open source, help to improve or make your fork and have fun, is how open source works right?

**Comment by doomedraven** [comment]
@ethhart yes that will be removed
